src/programs/bamsplitdiv.1
.TH BAMSPLITDIV 1 "October 2013" BIOBAMBAM
.SH NAME
bamsplitdiv - split BAM file into a set of BAM files
.SH SYNOPSIS
.PP
.B bamsplitdiv
[options]
.SH DESCRIPTION
bamsplitdiv reads a BAM file via standard input and creates a set of output
BAM files containing the alignments in the input. Given a positive natural
number 
.I div 
bamsplitdiv will create div files. The alignments in the input file will be
distributed to the output files in a round robin type scheme, i.e. entries
0, div, 2div, 3div, etc. will be written to the first output file, 1,div+1,2div+1,
3div+1, etc. to the second etc.
.PP
The following key=value pairs can be given:
.PP
.B div=<1>:
number of output files
.PP
.B prefix=<filename>
prefix for the create files. The created files will have names ${prefix}_000000.bam, ${prefix}_000001.bam, etc.
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B verbose=<1>:
Valid values are
.IP 1:
print progress report on standard error
.IP 0:
do not print progress report
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2013 German Tischler, \(co 2011-2013 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100

biobambam2 (2.0.22) unstable; urgency=medium

  * implicitly concatenate multiple input files as input for bamsort 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 29 Oct 2015 09:59:48 +0100

biobambam2 (2.0.21) unstable; urgency=medium

  * fix FF and RR type alignments in streaming duplicate marking
  * add MC tag (mate cigar string) when duplicate marking support is enabled in bamsort

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 26 Oct 2015 17:12:04 +0100

biobambam2 (2.0.20) unstable; urgency=medium

  * fix marking of FF and RR type duplicate records in BAM files 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 23 Oct 2015 14:55:46 +0200

biobambam2 (2.0.19) unstable; urgency=medium

  * include updates by Andrew Whitwham for improving the cigar handling of ClipAdapters

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 22 Sep 2015 16:51:25 +0200

biobambam2 (2.0.18) unstable; urgency=medium

  * update to libmaus2 2.0.68 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 14 Sep 2015 14:03:00 +0200

biobambam2 (2.0.17) unstable; urgency=medium

  * allow building without requirement for SSE4 (change in libmaus2) 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 03 Sep 2015 10:39:07 +0200

biobambam2 (2.0.16) unstable; urgency=medium

  * merge Andrew Whitwham's patch for removing adapter clipping tags from secondary and supplementary alignments 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 24 Aug 2015 13:15:02 +0200

biobambam2 (2.0.15) unstable; urgency=medium

  * update to libmaus2 2.0.45 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 21 Aug 2015 10:13:18 +0200

biobambam2 (2.0.14) unstable; urgency=medium

  * update to libmaus2 2.0.42 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 09 Aug 2015 13:36:32 +0200

biobambam2 (2.0.13) unstable; urgency=medium

  * allow in memory tmp files in bammarkduplicates, bammarkduplicates2 and bamtofastq 

 -- German Tischler <gt1@sanger.ac.uk>  Sat, 08 Aug 2015 15:03:19 +0200

biobambam2 (2.0.12) unstable; urgency=medium

  * update to libmaus2 2.0.40

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 07 Aug 2015 08:36:53 +0200

biobambam2 (2.0.11) unstable; urgency=medium

  * update to libmaus2 2.0.39 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 06 Aug 2015 18:23:03 +0200

biobambam2 (2.0.10) unstable; urgency=medium

  * include Andrew Whitwham's (streaming) duplicate marking tag changes

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 22 Jun 2015 11:22:34 +0200

biobambam2 (2.0.9) unstable; urgency=medium

  * add daligner las to SAM/BAM/CRAM conversion tool 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 19 Jun 2015 15:45:53 +0200

biobambam2 (2.0.8) unstable; urgency=low

  * fix for random failures in bamsormadup due to missing/incomplete initialisation 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 23:49:56 +0100

biobambam2 (2.0.7) unstable; urgency=low

  * update libmaus version to 2.0.6 (fix for bam indexing) 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 11:43:45 +0100

biobambam2 (2.0.6) unstable; urgency=low

  * update libmaus version to 2.0.5 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 24 Apr 2015 09:31:27 +0100

biobambam2 (2.0.5) unstable; urgency=low

  * fix/clarify manual page for bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 15:33:14 +0100

biobambam2 (2.0.4) unstable; urgency=low

  * add option to store temp data in RAM instead of on disc in bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 13:53:51 +0100

biobambam2 (2.0.3) unstable; urgency=medium

  * update to libmaus 2.0.3 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 12:48:32 +0100

biobambam2 (2.0.2) unstable; urgency=medium

  * update to libmaus 2.0.2 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 10:36:05 +0100

biobambam2 (2.0.1) unstable; urgency=medium

  * no change version bump to check version bumping script 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 00:15:29 +0100

biobambam2 (2.0.0) unstable; urgency=low

  * import of biobambam code

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 18 Apr 2015 12:00:00 +0100


src/programs/bamsplitdiv.cpp
/**
    bambam
    Copyright (C) 2009-2013 German Tischler
    Copyright (C) 2011-2013 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include "config.h"

#include <iostream>
#include <queue>

#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BamCat.hpp>
#include <libmaus2/bambam/BamWriter.hpp>

#include <biobambam2/Licensing.hpp>

static int getDefaultLevel() { return Z_DEFAULT_COMPRESSION; }
static int getDefaultVerbose() { return 1; }
static uint64_t getDefaultDiv() { return 1; }
static std::string getDefaultFilePrefix(::libmaus2::util::ArgInfo const & arginfo) { return arginfo.getDefaultTmpFileName(); }

::libmaus2::bambam::BamHeader::unique_ptr_type updateHeader(
	::libmaus2::util::ArgInfo const & arginfo,
	::libmaus2::bambam::BamHeader const & header
)
{
	std::string const headertext(header.text);

	// add PG line to header
	std::string const upheadtext = ::libmaus2::bambam::ProgramHeaderLineSet::addProgramLine(
		headertext,
		"bamsplitmod", // ID
		"bamsplitmod", // PN
		arginfo.commandline, // CL
		::libmaus2::bambam::ProgramHeaderLineSet(headertext).getLastIdInChain(), // PP
		std::string(PACKAGE_VERSION) // VN			
	);
	// construct new header
	::libmaus2::bambam::BamHeader::unique_ptr_type uphead(new ::libmaus2::bambam::BamHeader(upheadtext));
	
	return UNIQUE_PTR_MOVE(uphead);
}

int bamsplitmod(libmaus2::util::ArgInfo const & arginfo)
{
	if ( isatty(STDIN_FILENO) )
	{
		::libmaus2::exception::LibMausException se;
		se.getStream() << "Refusing read binary data from terminal, please redirect standard input to pipe or file." << std::endl;
		se.finish();
		throw se;
	}

	int const level = libmaus2::bambam::BamBlockWriterBaseFactory::checkCompressionLevel(arginfo.getValue<int>("level",getDefaultLevel()));
	int const verbose = arginfo.getValue<int>("verbose",getDefaultVerbose());
	uint64_t const div = arginfo.getValue<int>("div",getDefaultDiv());
	std::string const prefix = arginfo.getUnparsedValue("prefix",getDefaultFilePrefix(arginfo));
	
	if ( ! div )
	{
		::libmaus2::exception::LibMausException se;
		se.getStream() << "div cannot be 0." << std::endl;
		se.finish();
		throw se;
	}

	libmaus2::bambam::BamDecoder bamdec(std::cin);
	libmaus2::bambam::BamAlignment const & algn = bamdec.getAlignment();
	libmaus2::bambam::BamHeader const & header = bamdec.getHeader();
	::libmaus2::bambam::BamHeader::unique_ptr_type uphead(updateHeader(arginfo,header));
	
	libmaus2::autoarray::AutoArray<libmaus2::aio::OutputStreamInstance::unique_ptr_type> COS(div);
	libmaus2::autoarray::AutoArray<libmaus2::bambam::BamWriter::unique_ptr_type> writers(div);
	std::vector < std::string > filenames;
	for ( uint64_t i = 0; i < div; ++i )
	{
		std::ostringstream ostr;
		ostr << prefix << "_" << std::setw(6) << std::setfill('0') << i << std::setw(0) << ".bam";
	
		libmaus2::aio::OutputStreamInstance::unique_ptr_type tCOS(new libmaus2::aio::OutputStreamInstance(ostr.str()));
		COS[i] = UNIQUE_PTR_MOVE(tCOS);
		libmaus2::bambam::BamWriter::unique_ptr_type twriter(new libmaus2::bambam::BamWriter(*COS[i],*uphead,level));
		writers[i] = UNIQUE_PTR_MOVE(twriter);
	}

	uint64_t c = 0;
	if ( verbose )
	{
		while ( bamdec.readAlignment() )
		{
			algn.serialise ( writers [ (c++) % div ] -> getStream() );
			
			if ( ((c) & ((1ull<<20)-1)) == 0 )
				std::cerr << "[V] " << c << std::endl;
		}
		std::cerr << "[V] " << c << std::endl;
	}
	else
	{
		while ( bamdec.readAlignment() )
			algn.serialise ( writers [ (c++) % div ] -> getStream() );
	}

	for ( uint64_t i = 0; i < div; ++i )		
	{
		writers[i].reset();
		COS[i]->flush();
		COS[i].reset();
	}
	
	return EXIT_SUCCESS;
}

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);
		
		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if ( 
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if ( 
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;
				
				std::vector< std::pair<std::string,std::string> > V;
			
				V.push_back ( std::pair<std::string,std::string> ( "div=<["+::biobambam2::Licensing::formatNumber(getDefaultDiv())+"]>", "divisor" ) );
				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( "verbose=<["+::biobambam2::Licensing::formatNumber(getDefaultVerbose())+"]>", "print progress report" ) );
				V.push_back ( std::pair<std::string,std::string> ( "prefix=<["+getDefaultFilePrefix(arginfo)+"]>", "default output file prefix" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}
			
		return bamsplitmod(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}

